internal/xds/translator/ratelimit.go

```diff
@@ -401,25 +384,36 @@ func buildRateLimitServiceDescriptors(global *ir.GlobalRateLimit) []*rlsconfv3.R
 			pbDesc := new(rlsconfv3.RateLimitDescriptor)
 			pbDesc.Key = "masked_remote_address"
 			pbDesc.Value = rule.CIDRMatch.CIDR
-			rateLimit := rlsconfv3.RateLimitPolicy{
-				RequestsPerUnit: uint32(rule.Limit.Requests),
-				Unit:            rlsconfv3.RateLimitUnit(rlsconfv3.RateLimitUnit_value[strings.ToUpper(string(rule.Limit.Unit))]),
+
+			if cur != nil {
+				cur.Descriptors = []*rlsconfv3.RateLimitDescriptor{pbDesc}
+				cur = pbDesc
+			} else {
+				head = pbDesc
+				cur = head
 			}
 
 			if rule.CIDRMatch.Distinct {
-				pbDesc.Descriptors = []*rlsconfv3.RateLimitDescriptor{
-					{
-						Key:       "remote_address",
-						RateLimit: &rateLimit,
-					},
-				}
-			} else {
-				pbDesc.RateLimit = &rateLimit
+				pbDesc := new(rlsconfv3.RateLimitDescriptor)
+				pbDesc.Key = "remote_address"
+				cur.Descriptors = []*rlsconfv3.RateLimitDescriptor{pbDesc}
+				cur = pbDesc
 			}
+		}
+
+		// Case when header/cidr match is not set and the rate limit is applied
+		// to all traffic.
+		if !rule.IsMatchSet() {
+			pbDesc := new(rlsconfv3.RateLimitDescriptor)
+			// GenericKey case
+			pbDesc.Key = getRouteRuleDescriptor(rIdx, -1)
+			pbDesc.Value = getRouteRuleDescriptor(rIdx, -1)
 			head = pbDesc
 			cur = head
 		}
 
+		// Add the ratelimit values to the last descriptor
```

Comment:
this used to be under `if mIdx == len(rule.HeaderMatches)-1` , was that a bug ?

Reply:
yes, if we set both header and cidr matches, the `rateLimitPolicy` should not be added just because `mIdx == len(rule.HeaderMatches)-1`, it will be rewrite/coverd by the follow cidr match descriptors, causing header match to be invalid.

so the condition should be `mIdx == len(rule.HeaderMatches)-1 && no cidr match exist`

Reply:
the fix here is to add `rateLimitPolicy` after all the descriptors were built, works for: 1) only header match, 2) only cidr match, 3) both header & cidr match, 4) no match